### PR TITLE
Major refactoring :D

### PR DESCRIPTION
**Pull request by jcpsantiago**
Split logic into purer functions;
Pre-calculate more elements when package loads;
Use more of the Sentry SDK.

### CODE REVIEW

File: R/core.R [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    stacktrace = list(
+      frames = stacktrace
+    )
+  )
+
+  capture(exception = exception_payload, ..., level = level)
+}
+
+
+#' Build the response URL
```
Comment (kirel) [inline R/core.R]: Response?

File: R/core.R [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+#' @return a character string
+sentry_url <- function() {
+  sprintf(
+    "%s://%s/api/%s/store/",
+    .sentry_env$protocol, .sentry_env$host, .sentry_env$project_id
+  )
+}
+
+
+#' Set the response header
```
Comment (kirel) [inline R/core.R]: Response?

File: R/plumber_handler.R [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+#' Default error handler for Plumber
+#'
+#' @param req a Plumber request object
+#' @param res a Plumber response object
+#' @param error an error object
+#'
+#' @return a list
+#' @export
+default_error_handler <- function(req, res, error) {
+  print(error)
```
Comment (kirel) [inline R/plumber_handler.R]: Should this be here?

File: README.md [hunk 4]
```diff
@@ -5,49 +5,110 @@
...
+```r
+library(plumber)
+library(sentryR)
+
+configure_sentry(dsn = Sys.getenv('SENTRY_DSN'), 
+                 app_name = "myapp", app_version = "1.0.0")
 
 r <- plumb("R/api.R")
-r$setErrorHandler(errorHandler)
+r$setErrorHandler(sentry_error_handler)
```
Comment (kirel) [inline README.md]: Should this explain how to wrap the default error handler (or your own) instead?

File: tests/testthat/test-sentry.R [hunk 5]
```diff
@@ -1,86 +1,149 @@
...
+    parse_dsn(test_dsn),
+    list(
+      dsn = test_dsn, protocol = "https", public_key = "1234",
+      ignore = NA_character_, secret_key = NA_character_,
+      host = "sentry.io", project_id = "1"
+    )
+  )
+})
+
 test_that("setting configuration works", {
```
Comment (kirel) [inline tests/testthat/test-sentry.R]: Only tests what doesn't work - not what works?

File: tests/testthat/test-sentry.R [hunk 6]
```diff
@@ -1,86 +1,149 @@
...
+    function(x) exists(x, envir = .sentry_env)
+  )
 
   expect_true(all(fields))
 
-  rm(list = ls(envir = .SentryEnv), envir = .SentryEnv)
+  rm(list = ls(envir = .sentry_env), envir = .sentry_env)
 })
 
 test_that("we build the correct response url", {
```
Comment (kirel) [inline tests/testthat/test-sentry.R]: Request?

File: R/calls.R [hunk 7]
```diff
@@ -1,45 +1,111 @@
...
+        if (!is.null(file$original)) {
+          return(file$original$lines[(line + 1):(line + 5)])
+        }
+
+        return(file$lines[(line + 1):(line + 5)])
+      }
+      return(NA_character_)
+    }, srcfiles, lineno)
+  )
+
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/calls.R]: yeah, that's the only reason.. I love `purrr` but it's a large dependency

File: R/core.R [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+  }
+
+  # rm_null_obs transforms everything into a list, and we need
+  # the stacktrace as a data.frame/tibble
+  without_nulls$exception$stacktrace <- system_parameters$exception$stacktrace
+
+  payload <- jsonlite::toJSON(
+    without_nulls,
+    auto_unbox = TRUE,
+    null = "null",
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/core.R]: true, that line can go

File: R/core.R [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+#'
+#' @return sends message to Sentry
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' capture(message = "oh hai there!") # send message to sentry
+#' }
+capture <- function(...) {
+  browser()
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/core.R]: "surprise feature"

File: R/core.R [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+  packages <- as.list(versions)
+
+  sys_info <- Sys.info()
+
+  system_parameters <- list(
+    logger = "R",
+    platform = "R", # Sentry will ignore this for now
+    sdk = list(
+      name = "SentryR",
+      version = packages$sentryR
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/core.R]: calling `utils::installed.packages()` or even `utils::installed.packages()` when the package loads is not compatible with CRAN, that's why I'm doing the above. For that reason I prefer to keep the above solution for now, until we find a way around CRAN.

File: R/core.R [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+  resp_status_code <- httr::status_code(resp)
+
+  if (resp_status_code == 201 || resp_status_code == 200) {
+    resp_body <- httr::content(resp)
+    message("Your event was recorder in Sentry with ID ", resp_body$id)
+  } else {
+    warning(
+      "Error connecting to Sentry:",
+      httr::content(resp, "text", encoding = "UTF-8")
+    )
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/core.R]: If sentryR is being used within an API using `futures` that would work, but starting `futures` just for sentryR is overkill :(

File: R/core.R (hunk 1 above)
Comment (jcpsantiago) [reply R/core.R]: @kirel what do you think about `Build the sentry.io call URL`?

File: R/calls.R [hunk 12]
```diff
@@ -1,45 +1,111 @@
...
+      }
+      return(NA_character_)
+    }, srcfiles, lineno),
+    post_context = mapply(function(file, line) {
+      if (!is.null(file)) {
+        if (!is.null(file$original)) {
+          return(file$original$lines[(line + 1):(line + 5)])
+        }
+
+        return(file$lines[(line + 1):(line + 5)])
```
[the remark this answers is not in the document]
Comment (jcpsantiago) [reply R/calls.R]: Yes, it just returns the lines that actually exist.

**Pull request merged**